src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs

```diff
@@ -98,33 +105,90 @@ public static Checksum Create(IObjectWritable @object)
 
         public static Checksum Create(Checksum checksum1, Checksum checksum2)
         {
-            using var stream = SerializableBytes.CreateWritableStream();
+#if NET
+            return CreateUsingSpans(checksum1, checksum2);
+#else
+            return CrateUsingByteArrays(checksum1, checksum2);
+#endif
+        }
 
-            using (var writer = new ObjectWriter(stream, leaveOpen: true))
-            {
-                checksum1.WriteTo(writer);
-                checksum2.WriteTo(writer);
-            }
+        public static Checksum Create(Checksum checksum1, Checksum checksum2, Checksum checksum3)
+        {
+#if NET
+            return CreateUsingSpans(checksum1, checksum2, checksum3);
+#else
+            return CrateUsingByteArrays(checksum1, checksum2, checksum3);
+#endif
+        }
 
-            stream.Position = 0;
-            return Create(stream);
+        private static Checksum CrateUsingByteArrays(Checksum checksum1, Checksum checksum2)
+        {
+            using var hash = s_incrementalHashPool.GetPooledObject();
+            using var bytes = s_twoChecksumByteArrayPool.GetPooledObject();
+
+            var bytesSpan = bytes.Object.AsSpan();
+            checksum1.WriteTo(bytesSpan);
+            checksum2.WriteTo(bytesSpan.Slice(HashSize));
+
+            hash.Object.AppendData(bytes.Object);
+
+            return From(hash.Object.GetHashAndReset());
         }
 
-        public static Checksum Create(Checksum checksum1, Checksum checksum2, Checksum checksum3)
+        private static Checksum CrateUsingByteArrays(Checksum checksum1, Checksum checksum2, Checksum checksum3)
         {
-            using var stream = SerializableBytes.CreateWritableStream();
+            using var hash = s_incrementalHashPool.GetPooledObject();
+            using var bytes = s_threeChecksumByteArrayPool.GetPooledObject();
 
-            using (var writer = new ObjectWriter(stream, leaveOpen: true))
-            {
-                checksum1.WriteTo(writer);
-                checksum2.WriteTo(writer);
-                checksum3.WriteTo(writer);
-            }
+            var bytesSpan = bytes.Object.AsSpan();
+            checksum1.WriteTo(bytesSpan);
+            checksum2.WriteTo(bytesSpan.Slice(HashSize));
+            checksum3.WriteTo(bytesSpan.Slice(2 * HashSize));
 
-            stream.Position = 0;
-            return Create(stream);
+            hash.Object.AppendData(bytes.Object);
+
+            return From(hash.Object.GetHashAndReset());
         }
 
+#if NET
+
+        // Optimized helpers that do not need to allocate any arrays to combine hashes.
+
+        private static Checksum CreateUsingSpans(Checksum checksum1, Checksum checksum2)
+        {
+            using var hash = s_incrementalHashPool.GetPooledObject();
+
+            Span<byte> bytesSpan = stackalloc byte[2 * HashSize];
+            Span<byte> hashResultSpan = stackalloc byte[hash.Object.HashLengthInBytes];
```

Comment:
>Span<byte> hashResultSpan = stackalloc byte[hash.Object.HashLengthInBytes]; [](http://example.com/codeflow?start=12&length=75)

Is it worth the complexity to have the separate implementation for these? I feel like the pooled object one will have very similar performance numbers amortized.

Reply:
I'm amenable to either approach.  But this all seems self contained and nice to not even hit the heap.  Basically, we have a few 20 byte numbers and we're producing a final 20 byte number. Why not just have that work like normal math would work with ints on the stack:-)

Reply:
I'm missing something obvious, what's the reason the bytearray methods can't use stackalloc? I get that they can't use the new GetHashAndReset, I'm just wondering why they differ in the byte array allocation code.

Reply:
thanks for the explanation, haven't used stackalloc much and that helps